### ISSUE TITLE
Improve Project Structure

### ISSUE THREAD

**Issue by ilexp** [AdamsLair/duality#393]
### Summary

The current project folder structure in this repository has grown organically over the years and isn't ideal for new users. It should be re-organized.
### Analysis
- This needs to happen before [branching off v3.0](https://github.com/AdamsLair/duality/milestone/6), as it will make merges pretty much impossible.
- It will invalidate all previous repository links from the forum, since they don't use versioned links, but head links.
  - [Using commit-links](http://forum.adamslair.net/viewtopic.php?f=4&t=934) should prevent this in the future.
- It will also screw up file history, since git usually stops tracking after a rename / move.
  - Investigate whether it would make sense to [perform renames and move operations with git directly](https://git-scm.com/docs/git-mv).
  - Potentially not though, as the git move seems to be [a shortcut](https://jira.atlassian.com/browse/SRCTREEWIN-821?focusedCommentId=496696&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-496696) for a regular move with git add and git remove.
- Probably still worth it, as a long-term investment.
- Reduce top-level folders to a categorizing minimum.
  - **Source**: Contains the actual source code for all of Duality itself, as well as standard plugins.
  - **Build**: Contains build tool binaries, scripts and specs for building NuGet packages, Documentation, etc.
  - **Tools**: Containts source code for build tools, VS debugger integration, etc.
  - **.github**: GitHub-specific config files.
  - That's it? All else should be the regular and / or required top-level files.
- Every top-level folder except "Build" should be self-contained, so deleting all others will leave them unaffected.
- The "Source" folder should also be restructured internally:
  - Core (including primitives and tests)
  - Editor (including updater and tests)
  - Launcher
  - Plugins, Plugins/Tilemaps, Plugins/EditorModules, ...
  - Platform, Platform/Net45, ...
  - Samples, ...
  - Remove secondary solutions files and get back to the single "Duality.sln" one that contains everything. Make sure its internal folder structure matches the actual folder structure.
- Update `appveyor.yml`, potentially scripts.
- Update nightly build config file (relative paths).
- Do a test-merge into release to see if AppVeyor can do a complete build.
- Update Wiki pages with regard to source code links (Find / Replace All?)

**Comment by ChristianGreiner** [comment]
Will the changes affect to my current projects?

**Comment by ilexp** [comment]
> Will the changes affect to my current projects?

No. Regular Duality projects using the binaries are not affected at all. However, if you forked Duality and made custom changes to the repository, there may be some merging to do.

**Comment by ilexp** [comment]
This issue is on hold until [Hacktoberfest](https://hacktoberfest.digitalocean.com/) is over to avoid screwing up people's branches and PRs.

**Comment by ilexp** [comment]
Nice idea from [a folder structure conventions article](https://github.com/kriasoft/Folder-Structure-Conventions): Separate tests from program code. Top level folder structure would be:

- **Source**: Contains the actual source code for all of Duality itself, as well as standard plugins.
- **Test**: Contains all the tests, clearly separated from the actual code.
- **Build**: Contains build tool binaries, scripts and specs for building NuGet packages, Documentation, etc.
- **Tools**: Containts source code for build tools, VS debugger integration, etc.
- **.github**: GitHub-specific config files.

The internal folder structure of the Test folder could mimic the one from Source.

**Comment by ilexp** [comment]
### Progress

- Created a new project-structure-wip branch to work on this.
- Introduced the project structure that was detailed above.
- Fixed main solution file and projects
- Fixed tools solution file and projects
- Cleaned up and updated `.gitignore`
- Fixed NuSpecs
- Fixed NightlyBuild config
- Fixed Sandcastle project file
- Fixed `appveyor.yml`
- Fixed build scripts
- Fixed unit test project

### Immediate ToDo

- Evaluate the new structure and make sure it's better than before and worth merging.
- Merge back to `master`
- Merge to `release`
- Fix `appveyor.yml` in the `release` branch.
- Bump all package versions and do a full re-release of all packages. This serves the purpose of completely testing the new setup and also updating package icon URLs to match the new folder structure.

**Comment by ilexp** [comment]
@ChatPion and @JerTH pointed out in the chat that it would make sense to move `Samples` out of `Source` and into the root folder.

- It's more visible to newbies out there.
- It's not actually part of what one would consider the engine / editor sources, not even official plugin sources.

One thing to consider is that moving it out, they would be the second kind of project (besides tests) where the solution file would "link upwards" in folder structure. Which isn't great, but maybe not that bad either.

**Comment by ilexp** [comment]
### Progress

- Moved `Samples` from `Source` to root.
- Moved `Duality.sln` to root to account for the fact that it now contains projects from two other folders than `Source` only. Also makes it more visible as a starting point for everyone who wants to compile Duality.
- Fixed some more paths that were missed before.

### Immediate ToDo

- Evaluate the new structure and make sure it's good before merging.
- Branch the wiki repo and fix all source links. (Search for `/master/`)
- Merge back to `master` (both main and wiki)
- Merge to `release`
- Fix `appveyor.yml` in the `release` branch.
- Bump all package versions and do a full re-release of all packages. This serves the purpose of completely testing the new setup and also updating package icon URLs to match the new folder structure.

**Comment by ilexp** [comment]
Done
